### PR TITLE
Fetch_data function refactor with <no internet connection> bug fixed

### PR DESCRIPTION
**Pull request by TheKriserOne**

### CODE REVIEW

File: src/ui.rs [hunk 1]
```diff
@@ -2,28 +2,32 @@ use std::any::Any;
 use std::cell::RefCell;
 use std::rc::Rc;
 use std::str::FromStr;
+use secure_string::SecureString;
 use slint::ComponentHandle;
 use crate::auth::{AuthStore, Platform};
-use crate::{fetch_data, ChatGlobal, Conversation, MainWindow, SignInGlobal};
+use crate::{fetch_data, ChatGlobal, Conversation, MainWindow, Page, SignInGlobal};
```
Comment (laycookie) [inline src/ui.rs:8]: Page import is not needed

File: src/ui.rs [hunk 2]
```diff
@@ -2,28 +2,32 @@ use std::any::Any;
...
 use crate::backend::Messenger;
```
Comment (laycookie) [inline src/ui.rs:9]: Not needed here anymore

File: src/main.rs [hunk 3]
```diff
@@ -39,45 +43,36 @@ fn main() {
...
+        let mut auth_store = (*auth_store).borrow_mut();
+        let mut auths_to_remove = vec![];
+        for (i, auth) in auth_store.iter_mut().enumerate() {
+            match fetch_data(auth.platform.clone(), auth.token.clone(), &ui) {
+                Ok(..) => (),
+                (Err(error)) if error.status() == StatusCode::Unauthorized => {
+                    eprintln!("Token expired");
+                    auths_to_remove.push(i);
+                }
+                _ => eprintln!("There has been an issue with internet connection"),
```
Comment (laycookie) [inline src/main.rs:55]: This has to be more generic, as there are more errors then just a connection issue that might result in this, I belive.

File: src/main.rs [hunk 4]
```diff
@@ -39,45 +43,36 @@ fn main() {
...
+            }
+        }
+        auths_to_remove.sort_by(|a, b| b.cmp(a));
+        auths_to_remove.iter().for_each(|i| auth_store.remove(*i));
```
Comment (laycookie) [inline src/main.rs:59]: This feels like too much logic to be in the main, try moving in somewhere else. For example allow `fetch_data` to work with a slice of an array.

**Review by laycookie — CHANGES_REQUESTED** [review]

File: src/main.rs [hunk 5]
```diff
@@ -1,9 +1,13 @@
 use std::{cell::RefCell, rc::Rc};
+use secure_string::SecureString;
 use auth::{AuthStore};
 use backend::Messenger;
 #[cfg(all(not(debug_assertions), unix))]
 use daemonize::Daemonize;
 use slint::ComponentHandle;
+use surf::StatusCode;
+use crate::auth::Platform;
+use crate::backend::discord::json_structs::Profile;
```
[the remark this answers is not in the document]
Comment (TheKriserOne) [reply src/main.rs:10]: I do?(except the profile)

File: src/auth.rs [hunk 6]
```diff
@@ -103,29 +107,14 @@ impl AuthStore {
...
-        let mut reader = BufReader::new(&self.file);
-        reader.seek(SeekFrom::Start(0)).unwrap();
-
-        let mut lines = reader.lines().collect::<Vec<_>>();
-        lines.remove(i).unwrap();
 
-        // Prefferably I should just be writing to a new file, and then
-        // just swap the files when I'm finished writing, but realisticly
-        // there is no point in this type of redundncy at this point in the
-        // project.
```
Comment (laycookie) [inline src/auth.rs:119]: This comment still applies

File: src/auth.rs [hunk 7]
```diff
@@ -103,29 +107,14 @@ impl AuthStore {
...
+    pub fn remove(&mut self) {
```
Comment (laycookie) [inline src/auth.rs:111]: At this point, this function should probably be renamed to `file_sync`, and I'm not sure it should stay public.

**Review by laycookie — CHANGES_REQUESTED** [review]

File: src/main.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (laycookie) [reply src/main.rs]: Yes, I was specifically talking about the `profile`

File: src/main.rs [hunk 8]
```diff
@@ -38,46 +41,36 @@ fn main() {
     let ui = MainWindow::new().unwrap();
 
     // === Sign in, if user has a token ===
-    if !(*auth_store).borrow().is_empty() {
-        fetch_data(&ui, &auth_store.clone());
+    if !auth_store.is_empty() {
+        auth_store.retain_and_rewrite(|auth| {
```
Comment (laycookie) [inline src/main.rs:45]: Result will never return anything other then an error, so `_ => {...` will never get triggred

**Review by laycookie — CHANGES_REQUESTED** [review]

File: src/main.rs (hunk 8 above)
Comment (TheKriserOne) [reply src/main.rs:45]: it does get triggered IF status code doesnt equal unathorized but I think i will just put if statements inside an error arm-binding

**Pull request closed without merging**